File: archguard/dashboard/_capabilities.py
"""What this instance can actually do, for the interface to ask before offering.

The AI features degrade rather than fail: without ``GEMINI_API_KEY`` the
Advisor and remediation plans return an error per request. That is honest but
badly timed -- someone types a question, waits for a round trip, and is told the
feature was never going to work. The interface can know before it offers the
control.

Deliberately not a health check. ``/ready`` answers "should this instance take
traffic", and the AI features being off is not a reason to take an instance out
of rotation; it is a reason to grey out two buttons.
"""

from __future__ import annotations

import os
from typing import Any

from archguard.llm.gemini import ModelCheck, fallback_model, primary_model


class _State:
    """Holds the boot probe's result, if one was run.

    A class attribute rather than a module global so recording a result does
    not need `global`, which ruff flags and which makes the lifecycle harder to
    follow than it needs to be.
    """

    model_check: ModelCheck | None = None


_state = _State()


def record_model_check(result: ModelCheck | None) -> None:
    """Remember what the startup probe found, so callers need not repeat it."""
    _state.model_check = result


def ai_status() -> dict[str, Any]:
    """Whether the AI features will work, and if not, why.

    The reason is written for whoever has to fix it: a missing key names the
    variable, and a bad model id names the id. "AI unavailable" on its own
    sends an operator to the wrong place -- usually their credentials, when the
    model name is what is wrong.
    """
    # Mock mode serves canned AI responses without touching the API, which is
    # how the browser tests exercise the Advisor and remediation paths. The
    # features genuinely work in that mode, so reporting them unavailable would
    # be false -- and would disable the very controls those tests drive.
    if os.environ.get("ARCHGUARD_MOCK_LLM") == "1":
        return {"available": True, "reason": "", "mocked": True}

    if not os.environ.get("GEMINI_API_KEY", "").strip():
        return {
            "available": False,
            "reason": (
                "GEMINI_API_KEY is not set, so the AI Advisor and AI "
                "remediation plans are disabled."
            ),
        }

    check = _state.model_check
    # `checked=False` means the probe did not run or could not reach the API.
    # That says nothing about the models, and must not disable a deployment
    # that works.
    if check is not None and check.checked and not check.ok:
        return {
            "available": False,
            "reason": (
                f"The configured model(s) {', '.join(check.missing)} are not "
                "offered by the API, so AI requests would fail."
            ),
        }

    return {
        "available": True,
        "reason": "",
        "primary_model": primary_model(),
        "fallback_model": fallback_model(),
    }

File: archguard/dashboard/static/js/capabilities.js
import { safeFetch } from './api.js';
import { sanitize } from './dom.js';


/**
 * Stop offering what this instance cannot do.
 *
 * The AI features degrade rather than fail loudly: without `GEMINI_API_KEY` the
 * Advisor and remediation plans return an error per request. Honest, but badly
 * timed -- someone types a question, waits for a round trip, and is told the
 * feature was never going to work. Worse when the key is set and the configured
 * model id is wrong, because the error then reads like a credential problem.
 *
 * Asked once at load. A failed request leaves the controls enabled: not knowing
 * whether AI works is not the same as knowing it does not, and disabling a
 * working feature because a status call failed would be the worse mistake.
 */

/** Controls that do nothing useful without a working model. */
const AI_CONTROLS = [
    'btn-advisor-ask',
    'advisor-question-input',
    'remediation-btn',
    'violations-remediation-btn',
];

export async function applyCapabilities() {
    const caps = await safeFetch('/api/v1/capabilities', null);
    if (!caps || !caps.ai) return;

    if (caps.ai.available) {
        setAiEnabled(true, '');
        return;
    }
    setAiEnabled(false, caps.ai.reason || 'AI features are not configured.');
}


export function setAiEnabled(enabled, reason) {
    for (const id of AI_CONTROLS) {
        const el = document.getElementById(id);
        if (!el) continue;
        el.disabled = !enabled;
        if (enabled) {
            el.removeAttribute('title');
            el.removeAttribute('aria-describedby');
        } else {
            // Both, because the two audiences read different things: a
            // pointer user gets the tooltip, a screen-reader user gets the
            // description. `disabled` alone says "no" without saying why.
            el.title = reason;
            el.setAttribute('aria-describedby', 'ai-unavailable-note');
        }
    }

    const note = document.getElementById('ai-unavailable-note');
    if (!note) return;
    if (enabled) {
        note.hidden = true;
        note.textContent = '';
    } else {
        note.hidden = false;
        note.innerHTML = sanitize(reason);
    }
}
